tsconfig-base.json

```diff
@@ -0,0 +1,38 @@
+// Base compiler options and all compilation entry points.
+
+{
+  "compilerOptions": {
+    "composite": true,
+    "emitDeclarationOnly": true,
+    "declarationMap": true,
+
+    "target": "es2020",
+    "module": "es2020",
+    "moduleResolution": "node",
+    "esModuleInterop": true,
+
+    "allowJs": true,
+    "checkJs": true,
+    "strict": true,
+    // TODO: remove the next line to be fully `strict`.
+    "useUnknownInCatchVariables": false,
+
+    // "listFiles": true,
+    // "noErrorTruncation": true,
+    "extendedDiagnostics": true,
+  },
+  // `references` is not inherited by extending tsconfigs. This is a list of all
+  // project references so `tsc --build tsconfig-base.json` can type check all files.
+  "references": [
+    {"path": "./"},
+    {"path": "./types/lhr/"},
+    {"path": "./report/"},
+    {"path": "./report/generator/"},
+    {"path": "./lighthouse-viewer/"},
+    {"path": "./lighthouse-treemap/"},
+    {"path": "./flow-report/"},
+  ],
```

Comment:
it's arguable this should be _the_ `tsconfig.json` since this sets up compilation of all of lighthouse and today's `tsconfig.json` be `tsconfig-core.json` or whatever. That way you'd be able to just do `tsc -b` (which will just get you the core `tsconfig.json` today). But it also feels weird to have `tsconfig.json` not touching real code, so 🤷

Reply:
I'm a little confused with the purpose of tsconfig.json at this point though, it's just to be referenced by the base for core?

My rough mental model would be something like

`tsconfig-base.json`, extended by all others, none of these references
`tsconfig-core.json` checks just core
`tsconfig-all.json` checks everything, has these references

We don't need to do exactly that, but is that reflecting the three needs we have at the root directory?

Reply:
> `tsconfig-base.json`, extended by all others, none of these references
> `tsconfig-core.json` checks just core
> `tsconfig-all.json` checks everything, has these references

So yes, thanks, I skipped my reasoning here. This is a very common split (more or less the [recommended structure](https://www.typescriptlang.org/docs/handbook/project-references.html#overall-structure)).

Those are exactly our needs, but I hated having _three_ files for it in the root. `references` aren't inherited by extending configs, and if `tsconfig-base.json` doesn't have anything in `includes`/`files`, it also doesn't use the `compilerOptions` itself, so we can just combine `tsconfig-base.json` and `tsconfig-all.json` from your list and it functions the same exact way.

But if that's just more confusing than having an extra file, we can just have the extra file :)

Reply:
ok, all good now @patrickhulce. Let me know what you think about 

> But if that's just more confusing than having an extra file, we can just have the extra file

Reply:
> if tsconfig-base.json doesn't have anything in includes/files, it also doesn't use the compilerOptions itself

I followed everything else but this. Why does it need to? It's only purpose from my perspective is to be the thing from which others inherit :) My *expectation* is that such a base file does not check anything in particular, just sets up the options.

Reply:
Ah crossed paths in replies :) 

> But if that's just more confusing than having an extra file, we can just have the extra file

Yeah I guess that's the side I'm on, 1 -> 2 feels like the rough part, 2 -> 3 is more like 🤷  and a win for clarity IMO

Reply:
> Yeah I guess that's the side I'm on, 1 -> 2 feels like the rough part, 2 -> 3 is more like 🤷 and a win for clarity IMO

sounds good. I like that it's more straightforward and we'll just have to expand our tsconfig zoo to make room :)